Add support for Postgres 16

chuttam opened the pull request:
(Creating PR from a fork since I don't have necessary permissions to push branches/PRs on the base repo)

Some EasyRedir applications run the `3.1-14` image in which `pg_dump` and other client binaries are compatible with Postgres server 14. Now that applications are bumped to connect with databases running Postgres 16, the client-side libraries should be updated too.

For example, seeing compatibility issues when attempting to generate `structure.sql` on containers running `3.1-14` connected to postgres 16 database containers.

On Rails app server startup, or `rails db:setup`, etc.

```
pg_dump: error: server version: 16.1; pg_dump version: 14.8 (Ubuntu 14.8-1.pgdg22.04+1)
pg_dump: error: aborting because of server version mismatch
rake aborted!
failed to execute:
pg_dump --schema-only --no-privileges --no-owner --file /app/db/structure.sql er-management_dev

Please check the output above for any errors and make sure that `pg_dump` is installed in your PATH and has proper permissions.


```

chuttam committed:
Add support for Postgres 16

wgrrrr commented:
This is great, thank you! If you want to fully complete this task, you will need to make an edit to `.github/workflows/docker-image.yml` and include a new block of code to tell Github Actions to create and push the new image to Docker Hub. For example:

```yaml
    - name: Build and push Ruby 3.1 with PostgreSQL 16 client
      uses: docker/build-push-action@v2
      with:
        context: .
        file: ./3.1-16/Dockerfile
        platforms: linux/amd64,linux/arm64
        push: true
        tags: easyredir/ruby-psql:3.1-16
        cache-from: type=registry,ref=user/app:latest
        cache-to: type=inline
```

Building these images can take a half hour to an hour per image. Given that, I typically do some house cleaning when adding a new image to _also_ remove old images which are no longer needed, which keeps the build times down.

I think we can remove the blocks for all Ruby versions prior to 3.1. We are not using them anywhere and they just make the build substantially longer.

chuttam committed:
Add 3.1 PG 16 image build to Github Actions

Also removes pre 3.0 ruby images.

chuttam committed:
Parallelize image build actions

TODO: Use a composite action to DRY up the initial setup for each image build.

chuttam commented:
@wgrrrr  I went ahead and removed the 2.x blocks (left the 3.0 one in there... 🤷 )

Also, I refactored the Actions manifest to parallel the 3 image builds. This introduces a bit of duplication in the common set up of each job, which we can DRY up with a "composite" action here soon.
